torch/csrc/jit/passes/reconstruct_scopes.cpp

```diff
@@ -0,0 +1,105 @@
+#include <torch/csrc/jit/passes/reconstruct_scopes.h>
+#include <torch/csrc/jit/jit_log.h>
+
+namespace torch {
+namespace jit {
+
+class ReconstructScopesPass {
+ public:
+  ReconstructScopesPass(const Module& m, Graph& g, std::string p)
+      : root_module(&m), graph(&g), prefix(std::move(p)) {};
+  void run();
+
+ private:
+  const Module* root_module;
+  Graph* graph;
+  std::string prefix;
+
+  std::unordered_map<Function*, ModulePtr> func_to_module;
+  std::unordered_map<ModulePtr, std::string> module_names;
+
+  void visitBlock(Block* b);
+  void visitNode(Node* n);
+
+  void constructFunctionToModuleMap(const Module& module);
+  void constructRelativeNamesForModules(
+      const Module& module,
+      const std::string& prefix);
+
+  std::string getScopeString(Function* f) const;
+};
+
+void ReconstructScopesPass::constructFunctionToModuleMap(const Module& module) {
+  for (auto& method : module.get_methods()) {
+    func_to_module[&method.function()] = module._ivalue();
+  }
+  for (Module m : module.children()) {
+    constructFunctionToModuleMap(m);
+  }
+}
+
+void ReconstructScopesPass::constructRelativeNamesForModules(
+    const Module& module,
+    const std::string& prefix) {
+  module_names[module._ivalue()] = prefix;
+  for (NameModule s : module.named_children()) {
+    constructRelativeNamesForModules(s.value, prefix + "." + s.name);
+  }
+}
+
+std::string ReconstructScopesPass::getScopeString(Function* f) const {
+  if (!func_to_module.count(f)) {
+    return "<null (no func in the map)>";
+  }
+  auto m = func_to_module.at(f);
+  if (!module_names.count(m)) {
+    return "<null (no module in the map)>";
+  }
+  return module_names.at(m) + "." + f->name();
+}
+
+void ReconstructScopesPass::visitNode(Node* n) {
```

Comment:
const Node

Reply:
I did try `const` with `Node*`, `Block*`, and `Function*`, but it raised compiler errors because we modified the objects such as `n->setScope(...)`

Reply:
Aah. Ok nevermind.